libraries/standard/mqtt/src/mqtt.c

```diff
@@ -948,14 +954,33 @@ static MQTTStatus_t receiveConnack( MQTTContext_t * const pContext,
                                                       pContext->transportInterface.networkContext,
                                                       pIncomingPacket );
 
-        /* Loop until there is data to read or if the timeout has not expired. */
-    } while( ( status == MQTTNoDataAvailable ) &&
-             ( calculateElapsedTime( getTimeStamp(), entryTimeMs ) < timeoutMs ) );
+        /* The loop times out based on 2 conditions.
+         * 1. If a non-NULL getTime function is provided by the application:
+         *    Loop times out based on the timeout calculated by getTime()
+         *    function.
+         * 2. If a non-NULL getTime function is not provided by the application:
+         *    Loop times out based on the maximum number of retries config
+         *    MQTT_CONNACK_RECEIVE_RETRY_COUNT. */
+        if( getTimeStamp != NULL )
+        {
+            /* Loop runs only once for 0 timeoutMs. */
+            breakFromLoop = ( bool ) ( calculateElapsedTime( getTimeStamp(), entryTimeMs ) >= timeoutMs );
+        }
+        else
+        {
+            breakFromLoop = ( bool ) ( loopCount++ >= MQTT_CONNACK_RECEIVE_RETRY_COUNT );
```

Comment:
Maybe add a comment clarifying that the maximum number of iterations this loop can run is `MQTT_CONNACK_RECEIVE_RETRY_COUNT + 1` (1 try + number of retries)

Reply:
I think the word retry speaks for itself. Anyway if you had to rethink, I think it makes sense to add comment regarding that.